server/src/main/java/io/envoyproxy/controlplane/server/DiscoveryRequestStreamObserver.java

```diff
@@ -155,7 +151,13 @@ private void send(Response response, String typeUrl) {
     // Store the latest response *before* we send the response. This ensures that by the time the request
     // is processed the map is guaranteed to be updated. Doing it afterwards leads to a race conditions
     // which may see the incoming request arrive before the map is updated, failing the nonce check erroneously.
-    setLatestResponse(typeUrl, discoveryResponse);
+    setLatestResponse(
+        typeUrl,
+        LatestDiscoveryResponse.create(
+            nonce,
+            resources.stream().map(Resources::getResourceName).collect(Collectors.toSet())
```

Comment:
I think we can further optimise it by changing this line to:
```java
response.resources().stream().map(Resources::getResourceName).collect(Collectors.toSet())
```
avoiding unpacking all resources this way